Add initial lifetime checker

pfultz2 opened the pull request:
This checks for dangling lifetimes with lambdas and containers. It can warn about issue like this:

```cpp
auto f(std::vector<int> a) {
    auto it = a.begin();
    return [=](){ return it; };
}
```

Or issues where the lambda captures local variables by reference:

```cpp
auto f() {
    int a = 1;
    return [&](){ return a; };
}
```

Right now it only tracks lifetimes of references, pointers, containers, iterators, and lambdas. I would like to extend it in the future to functions and constructors. 

It also only warns for dangling returns. I would like to extend it to find dangling lifetimes from variable usages.

danmar commented on test/testautovariables.cpp:
```diff
@@ -1183,6 +1188,161 @@ class TestAutoVariables : public TestFixture {
+              "    auto l = [&](){ return a; };\n"
+              "    return l;\n"
+              "}\n");
+        ASSERT_EQUALS("[test.cpp:2] -> [test.cpp:4]: (error) Returning object with dangling lifetime from local variable 'a'.\n", errout.str());
```
I want that we test `ValueFlow` in testvalueflow. so most of these tests seem to belong in that file.

The algorithm of the checker is basically just to lookup "return|throw". And then if the operand has a dangling lifetime according to valueflow there is a error. We can test the "return|throw" lookup here and how the valueflow is used. But not the valueflow analysis.

In testvalueflow it should be tested if the operand get the value properly.

I am not sure how it should be splitted... but there is no test in testvalueflow right now.

danmar commented on lib/valueflow.cpp:
```diff
@@ -2654,6 +2654,189 @@ static void valueFlowAfterCondition(TokenList *tokenlist, SymbolDatabase* symbol
+            arguments = capture->link()->next();
+        }
+        const Token * afterArguments = arguments ? arguments->link()->next() : capture->link()->next();
+        if(Token::simpleMatch(afterArguments, "->")) {
```
This condition is never true. The `->` operators are replaced with `.` in the tokenlist. For information, if you need to know if the original code says `->` then you can use `Token::originalName()`.

pfultz2 replied:
I see, I thought it was only replaced for member accesses.

danmar commented:
It seems to be a good checker!

pfultz2 commented on lib/astutils.cpp:
```diff
@@ -135,9 +135,18 @@ const Token * astIsVariableComparison(const Token *tok, const std::string &comp,
 {
-    if (!tok || !tok->astOperand1())
+    const Token * rightmostLeaf = tok;
+    if (!rightmostLeaf || !rightmostLeaf->astOperand1())
```
It doesn't find the end of the lambda expression. It causes this test to fail:

```cpp
auto f() {
    int a = 1;
    auto l = [&](){ return a; };
    return l;
}
```

I tried to update the original expressionString code but couldn't make it work. I can add tests for it in test/testastutils.cpp.

pfultz2 commented:
> I suggest that object is changed to lambda.

It works for more than just lambdas. It can warn for cases when we are returning a dangling iterator:

```cpp
auto f(std::vector<int> x) {
    auto it = x.begin();
    return it;
}
```

Also I would like to extend it to track lifetime through functions and object constructors. Maybe start with aggregate constructor since it easier to reason about:

```cpp
struct foo
{
    std::vector<int>::iterator it;
};

foo f(std::vector<int> x) {
    auto it = x.begin();
    return foo{it};
}
```

> The wording "dangling lifetime" .. that does not give me a very good feeling. however I have no better suggestions right now.

Yea I would like better wording too, but I can't think of one. It was inspired by Herb Sutter's [Lifetime proposal](https://herbsutter.com/2018/09/20/lifetime-profile-v1-0-posted/). Although in its implementation it always says `a dangling pointer` even when its an object(like `string_view`) which seems confusing to me.

Maybe it would be better to say something like `Returning an object where the lifetime from local variable 'a' will be invalidated when exiting scope`?

danmar commented:
> It works for more than just lambdas. It can warn for cases when we are returning a dangling iterator:

Imho it would be better to write lambda/iterator explicitly.

I was a bit confused by the message when I saw the code. Since I am not 100% familiar with lambdas I missed that there was no `()` to call the lambda. So I thought the the lambda was called.. and also I am not 100% familiar with the rules for lambdas so I thought; maybe the function returns a `int &`.. and that is why you warn.

So well I think it could help some non-experts like me to understand the message.

> Maybe it would be better to say something like Returning an object where the lifetime from local variable 'a' will be invalidated when exiting scope?

How about:

Returning a lambda that must be executed inside the function because it references local variables.

pfultz2 commented:
> Imho it would be better to write lambda/iterator explicitly.

I will need to add a field to `ValueFlow::Value` to store what type of object it is. 

> Returning a lambda that must be executed inside the function because it references local variables.

That seems to imply that the error is not calling the lambda which is most likely not the case since the compilers can already check the types(unless its a lambda that returns another lambda). The error is the user needs to change the way they are capturing the variables by either switching to capturing by value or capturing a different object all together.

For lambdas we could write:

    Returning lambda that captures local variable 'a' whose lifetime will be invalidated when exiting

For iterators:

    Returning iterator to container 'a' whose lifetime will be invalidated when exiting

For objects(when we support them in the future):

   Returning object that points to local variable 'a' whose lifetime will be invalidated when exiting

For pointers I am not sure the best way yet. Right now, I warn for cases like this:

```cpp
int * f() {
    std::vector<int> v;
    int * p = v.data();
    return p;
}
```

But I can't just say `Returning pointer to local variable 'v'` as its actually a pointer to the data of the container so it maybe confusing to the user. For analysis purposes its the same thing since the data of the container and the variable share the same lifetime. It would be nice to say something like `Returning pointer to either local variable 'v' or a member of local variable 'v'`.

danmar commented:
> That seems to imply that the error is not calling the lambda which is most likely not the case since the compilers can already check the types

Yes.

> Returning lambda that captures local variable 'a' whose lifetime will be invalidated when exiting

I like the first half. Writing the variable name is nice but not crucial imho. I would like that it explicitly says that it captures the variable _by reference_. If I am not mistaken there is no problem if it captures the variable by value?

I have some doubts about "whose lifetime will be invalidated when exiting". Is it the _lifetime_ that is invalidated? I would like to avoid saying "exiting" .. "returning" or "goes out of scope" would be better.

For me, I think it's ok that we leave out this last part. And just write `Returning lambda that captures local variable 'a' by reference.`.

pfultz2 commented:
> If I am not mistaken there is no problem if it captures the variable by value?

No there can still be a problem. Here are three cases we warn for when returning a lambda which captures by value:

```cpp
auto f1() {
    int a = 1;
    int * p = &a;
    return [=]() { return p; };
}

auto f2() {
    int a = 1;
    auto f = [&]() { return a; };
    return [=]() { return f(); };
}

auto f3() {
    std::vector<int> v;
    auto it = v.begin();
    return [=]() { return it; };
}
```

In the first two cases we could say we are capturing `a` indirectly by reference, however, in the last case we are not capturing `v` indirectly by reference.

I think I need to look at adding more to the error path especially for these cases.

> Is it the lifetime that is invalidated? As I see it, the variable itself is invalidated, and the lifetime ends when the variable goes out of scope. 

I see it as the variable ends and the lifetime is invalid, but I can change the wording to say variable instead of lifetime.

> I would like to avoid saying "exiting" .. "returning" or "goes out of scope" would be better.

Yea that sounds better.

> For me, I think it's ok that we leave out this last part. And just write Returning lambda that captures local variable 'a' by reference.. Do you have a better suggestion?

But this error message doesn't explain why capturing the local variable is a problem in the cases its not captured by reference.

danmar commented:
> No there can still be a problem. Here are three cases we warn for when returning a lambda which captures by value:

ok.

it's very cool that these problems will be detected.

> But this error message doesn't explain why capturing the local variable is a problem in the cases its not captured by reference.

no it does not. I would say it's a good message in this case and I think we will be able to agree about some good variations of the messages.

danmar commented:
I tried to fix the cert.py issue with b5241ecb806f365febebb937169ec1571fda5970 .. I hope Travis will be happy if you merge/rebase.

pfultz2 commented:
I think this is ready to be merged unless you have more changes you want.

pfultz2 commented:
I am not sure why it hangs on clang, perhaps it needs to be rebuilt.

pfultz2 commented:
So I fixed the hang. This was due to `findLambdaEndToken` traversing with the ast but in cases such as constructing arrays(ie `T{1, 2, 3}[3]`) it was traversing backwards. So I added extra checks to make sure the traversal moves forward and not backwards.

pfultz2 commented:
I think this is ready to be merged in now.
